File: RNNoise (macOS)/Audio Unit/RNNoise__macOS_AudioUnit.m
//
//  RNNoise__macOS_AudioUnit.m
//  RNNoise (macOS)
//
//  Created by William on 24 Dec.
//

#import "RNNoise__macOS_AudioUnit.h"

#import <AVFoundation/AVFoundation.h>

// Define parameter addresses.
const AudioUnitParameterID voiceConfidenceThresholdID = 0;
const AudioUnitParameterID noiseReductionVolumeID = 1;

@interface RNNoise__macOS_AudioUnit ()

@property (nonatomic, readwrite) AUParameterTree *parameterTree;
@property AUAudioUnitBusArray *inputBusArray;
@property AUAudioUnitBusArray *outputBusArray;
@end


@implementation RNNoise__macOS_AudioUnit
@synthesize parameterTree = _parameterTree;

- (instancetype)initWithComponentDescription:(AudioComponentDescription)componentDescription options:(AudioComponentInstantiationOptions)options error:(NSError **)outError {
    self = [super initWithComponentDescription:componentDescription options:options error:outError];
    
    if (self == nil) { return nil; }

	_kernelAdapter = [[RNNoise__macOS_DSPKernelAdapter alloc] init];

	self.maximumFramesToRender = _kernelAdapter.maximumFramesToRender;

	[self setupAudioBuses];
	[self setupParameterTree];
	[self setupParameterCallbacks];
    return self;
}

#pragma mark - AUAudioUnit Setup

- (void)setupAudioBuses {
	// Create the input and output bus arrays.
	_inputBusArray  = [[AUAudioUnitBusArray alloc] initWithAudioUnit:self
															 busType:AUAudioUnitBusTypeInput
															  busses: @[_kernelAdapter.inputBus]];
	_outputBusArray = [[AUAudioUnitBusArray alloc] initWithAudioUnit:self
															 busType:AUAudioUnitBusTypeOutput
															  busses: @[_kernelAdapter.outputBus]];
}

- (void)setupParameterTree {
    // Create parameter objects.
    AUParameter *voiceConfidenceThreshold = [AUParameterTree createParameterWithIdentifier:@"voiceConfidenceThreshold"
																	name:@"Voice Confidence Threshold"
																 address:voiceConfidenceThresholdID
																	 min:0
																	 max:100
																	unit:kAudioUnitParameterUnit_Percent
																unitName:nil
																   flags:0
															valueStrings:nil
													 dependentParameters:nil];
    AUParameter *noiseReductionVolume = [AUParameterTree createParameterWithIdentifier:@"noiseReductionVolume"
                                                                                  name:@"Noise Reduction Volume"
                                                                               address:noiseReductionVolumeID
                                                                                   min:0
                                                                                   max:100
                                                                                  unit:kAudioUnitParameterUnit_Percent
                                                                              unitName:nil
                                                                                 flags:0
                                                                          valueStrings:nil
                                                                   dependentParameters:nil];

    // Initialize the parameter values.
    voiceConfidenceThreshold.value = 0.95;
    noiseReductionVolume.value = 0.95;

    // Create the parameter tree.
    _parameterTree = [AUParameterTree createTreeWithChildren:@[ voiceConfidenceThreshold, noiseReductionVolume ]];
}

- (void)setupParameterCallbacks {
	// Make a local pointer to the kernel to avoid capturing self.
	__block RNNoise__macOS_DSPKernelAdapter * kernelAdapter = _kernelAdapter;

	// implementorValueObserver is called when a parameter changes value.
	_parameterTree.implementorValueObserver = ^(AUParameter *param, AUValue value) {
		[kernelAdapter setParameter:param value:value];
	};

	// implementorValueProvider is called when the value needs to be refreshed.
	_parameterTree.implementorValueProvider = ^(AUParameter *param) {
		return [kernelAdapter valueForParameter:param];
	};

	// A function to provide string representations of parameter values.
	_parameterTree.implementorStringFromValueCallback = ^(AUParameter *param, const AUValue *__nullable valuePtr) {
		AUValue value = valuePtr == nil ? param.value : *valuePtr;

		return [NSString stringWithFormat:@"%.f", value];
	};
}

#pragma mark - AUAudioUnit Overrides

// If an audio unit has input, an audio unit's audio input connection points.
// Subclassers must override this property getter and should return the same object every time.
// See sample code.
- (AUAudioUnitBusArray *)inputBusses {
	return _inputBusArray;
}

// An audio unit's audio output connection points.
// Subclassers must override this property getter and should return the same object every time.
// See sample code.
- (AUAudioUnitBusArray *)outputBusses {
	return _outputBusArray;
}

// Allocate resources required to render.
// Subclassers should call the superclass implementation.
- (BOOL)allocateRenderResourcesAndReturnError:(NSError **)outError {
	if (_kernelAdapter.outputBus.format.channelCount != _kernelAdapter.inputBus.format.channelCount) {
		if (outError) {
			*outError = [NSError errorWithDomain:NSOSStatusErrorDomain code:kAudioUnitErr_FailedInitialization userInfo:nil];
		}
		// Notify superclass that initialization was not successful
		self.renderResourcesAllocated = NO;

		return NO;
	}

	[super allocateRenderResourcesAndReturnError:outError];
	[_kernelAdapter allocateRenderResources];
	return YES;
}

// Deallocate resources allocated in allocateRenderResourcesAndReturnError:
// Subclassers should call the superclass implementation.
- (void)deallocateRenderResources {
	[_kernelAdapter deallocateRenderResources];

    // Deallocate your resources.
    [super deallocateRenderResources];
}

- (NSTimeInterval) latency {
    // This AU does no buffering (so far) and thus I'm going to pretend it has
    // no latency for the sake of simplicity.
    return 0;
}

- (NSTimeInterval) tailTime {
    // Tail time may actually be negative because this AU removes noise, but
    // again 0 for the sake of simplicity.
    return 0;
}

#pragma mark - AUAudioUnit (AUAudioUnitImplementation)

// Block which subclassers must provide to implement rendering.
- (AUInternalRenderBlock)internalRenderBlock {
	return _kernelAdapter.internalRenderBlock;
}

@end

File: RNNoise (macOS)/Audio Unit/RNNoise__macOS_AudioUnit.swift
//
//  RNNoise__macOS_AudioUnit.swift
//  RNNoise (macOS)
//
//  Created by William on 29 Dec.
//

import Foundation
import AudioToolbox
import AVFoundation
import CoreAudioKit

public class RNNoise__macOS_AudioUnit: AUAudioUnit {

    private let kernelAdapter: RNNoise__macOS_DSPKernelAdapter

    lazy private var inputBusArray: AUAudioUnitBusArray = {
        AUAudioUnitBusArray(audioUnit: self, busType: .input, busses: [kernelAdapter.inputBus])
    }()

    lazy private var outputBusArray: AUAudioUnitBusArray = {
        AUAudioUnitBusArray(audioUnit: self, busType: .output, busses: [kernelAdapter.outputBus])
    }()

    public override var inputBusses: AUAudioUnitBusArray {
        return inputBusArray
    }

    public override var outputBusses: AUAudioUnitBusArray {
        return outputBusArray
    }

    public override var parameterTree: AUParameterTree? {
        get { return AUParameterTree.createTree(withChildren: []) }
        set { /* no setting parameter tree */ }
    }

    private var _currentPreset: AUAudioUnitPreset?

    /// The currently selected preset.
    public override var currentPreset: AUAudioUnitPreset? {
        get { return _currentPreset }
        set {
            // If the newValue is nil, return.
            guard newValue != nil else {
                _currentPreset = nil
                return
            }

            _currentPreset = nil

//            // Factory presets need to always have a number >= 0.
//            if preset.number >= 0 {
//                let values = factoryPresetValues[preset.number]
//                parameters.setParameterValues(cutoff: values.cutoff, resonance: values.resonance)
//                _currentPreset = preset
//            }
//            // User presets are always negative.
//            else {
//                // Attempt to restore the archived state for this user preset.
//                do {
//                    fullStateForDocument = try presetState(for: preset)
//                    // Set the currentPreset after we've successfully restored the state.
//                    _currentPreset = preset
//                } catch {
//                    print("Unable to restore set for preset \(preset.name)")
//                }
//            }
        }
    }

    public override var supportsUserPresets: Bool {
        return false
    }

    public override init(componentDescription: AudioComponentDescription, options: AudioComponentInstantiationOptions = []) throws {
        // Create adapter to communicate with underlying C++ DSP code.
        kernelAdapter = RNNoise__macOS_DSPKernelAdapter()

        try super.init(componentDescription: componentDescription, options: options)

        // Log component description values
        log(componentDescription)
    }

    private func log(_ acd: AudioComponentDescription) {

        let info = ProcessInfo.processInfo
        print("\nProcess Name: \(info.processName) PID: \(info.processIdentifier)\n")

        let message = """
        RNNoise__macOS_ (
                  type: \(acd.componentType.stringValue)
               subtype: \(acd.componentSubType.stringValue)
          manufacturer: \(acd.componentManufacturer.stringValue)
                 flags: \(String(format: "%#010x", acd.componentFlags))
        )
        """
        print(message)
    }

    public override var maximumFramesToRender: AUAudioFrameCount {
        get {
            return kernelAdapter.maximumFramesToRender
        }
        set {
            if !renderResourcesAllocated {
                kernelAdapter.maximumFramesToRender = newValue
            }
        }
    }

    public override func allocateRenderResources() throws {
        if kernelAdapter.outputBus.format.channelCount != kernelAdapter.inputBus.format.channelCount {
            throw NSError(domain: NSOSStatusErrorDomain, code: Int(kAudioUnitErr_FailedInitialization), userInfo: nil)
        }
        try super.allocateRenderResources()
        kernelAdapter.allocateRenderResources()
    }

    public override func deallocateRenderResources() {
        super.deallocateRenderResources()
        kernelAdapter.deallocateRenderResources()
    }

    public override var internalRenderBlock: AUInternalRenderBlock {
        return kernelAdapter.internalRenderBlock()
    }

    public override var canProcessInPlace: Bool {
        return true
    }

    public override var latency: TimeInterval {
        // This AU does no buffering yet, thus it has no latency to report.
        return 0;
    }

    public override var tailTime: TimeInterval {
        // This AU does not do reverb-style effects, thus it has no tail time to report.
        return 0;
    }
}

extension FourCharCode {
    var stringValue: String {
        let value = CFSwapInt32BigToHost(self)
        let bytes = [0, 8, 16, 24].map { UInt8(value >> $0 & 0x000000FF) }
        guard let result = String(bytes: bytes, encoding: .utf8) else {
            return "fail"
        }
        return result
    }
}
